personal_graph/graph.py

```diff
@@ -1,199 +1,567 @@
+                else:
+                    fetched_node_id = self.db.fetch_node_id(rowid[0])
+
+                if fetched_node_id is None:
+                    continue
+
+                node_data = self.db.search_node(fetched_node_id[0])
+                node_label = self.db.search_node_label(fetched_node_id[0])
+
+                if node_data in similar_rows:
+                    continue
+                node = Node(
+                    id=node_data["id"], label=node_label[0], attributes=node_data
+                )
+                similar_rows.append(node)
+
+        return similar_rows
+
+    def visualize(self, file: str, id: List[str]) -> Digraph:
+        return self.db.graphviz_visualize(file, id)
 
     def fetch_ids_from_db(self) -> List[str]:
-        return atomic(nodes_list(), self.db_url, self.auth_token)
+        return self.db.fetch_ids_from_db()
 
-    def search_indegree_edges(self, target) -> List[Any]:
-        return atomic(find_indegree_edges(target), self.db_url, self.auth_token)
+    def search_indegree_edges(self, target: str) -> List[Any]:
+        return self.db.search_indegree_edges(target)
 
-    def search_outdegree_edges(self, source) -> List[Any]:
-        return atomic(find_outdegree_edges(source), self.db_url, self.auth_token)
+    def search_outdegree_edges(self, source: str) -> List[Any]:
+        return self.db.search_outdegree_edges(source)
 
-    def is_unique_prompt(self, text: str, threshold: float) -> bool:
-        similar_nodes = atomic(
-            vector_search_node({"body": text}, threshold, 1),
-            self.db_url,
-            self.auth_token,
-        )
+    def is_unique_prompt(self, text: str, *, threshold: Optional[float] = 0.9) -> bool:
+        similar_nodes = self._similarity_search_node(text, threshold=threshold, limit=1)
 
         if not similar_nodes:
             return True
+
         return False
+
+    def insert(
+        self,
+        text: str,
+        attributes: Dict,
+    ) -> None:
+        node = Node(
+            id=str(uuid.uuid4()),
+            label=text,
+            attributes=json.dumps(attributes),
+        )
+        self.add_node(node)
+
+    def search(
+        self,
+        text: str,
+        *,
+        threshold: float = None,
+        descending: bool = False,
+        limit: int = 1,
+        sort_by: str = "",
+    ):
+        try:
+            similar_nodes = self._similarity_search_node(
+                text,
+                threshold=threshold,
+                descending=descending,
+                limit=limit,
+                sort_by=sort_by,
+            )
+
+        except Exception as e:
+            return e
+
+        if similar_nodes is None:
+            return None
+
+        if isinstance(self.vector_store, VliteVSS):
+            if limit == 1:
+                return similar_nodes[0][2]
+        else:
+            if limit == 1:
+                return json.loads(similar_nodes[0][3])
```

Comment:
why 3? avoid magic numbers

Reply:
This will be used whenever limit is 1 and user do not want to print the entire list of Nodes.
As per Readme that you have updated, we need to return the exact value of the specific key, so we need to fetch that key from the attributes. That's why I made use of this.
Else if fetching exact key value pair is not required then this won't be needed anymore.